src/librustdoc/json/mod.rs

```diff
@@ -1,47 +1,204 @@
+        } else if let types::ItemEnum::EnumItem(ref mut e) = new_item.inner {
+            e.impls = self.get_impls(id, cache)
+        }
+        self.index.borrow_mut().insert(id.into(), new_item);
+    }
+
+    fn get_trait_implementors(
+        &self,
+        id: rustc_span::def_id::DefId,
+        cache: &Cache,
+    ) -> Vec<types::Id> {
+        cache
+            .implementors
+            .get(&id)
+            .map(|implementors| {
+                implementors
+                    .iter()
+                    .map(|i| {
+                        let item = &i.impl_item;
+                        self.insert(item.clone(), cache);
+                        item.def_id.into()
+                    })
+                    .collect()
+            })
+            .unwrap_or_default()
+    }
+
+    fn get_impls(&self, id: rustc_span::def_id::DefId, cache: &Cache) -> Vec<types::Id> {
+        cache
+            .impls
+            .get(&id)
+            .map(|impls| {
+                impls
+                    .iter()
+                    .filter_map(|i| {
+                        let item = &i.impl_item;
+                        if item.def_id.is_local() {
+                            self.insert(item.clone(), cache);
+                            Some(item.def_id.into())
+                        } else {
+                            None
+                        }
+                    })
+                    .collect()
+            })
+            .unwrap_or_default()
+    }
+}
 
 impl FormatRenderer for JsonRenderer {
     fn init(
-        _krate: clean::Crate,
+        krate: clean::Crate,
         _options: RenderOptions,
         _render_info: RenderInfo,
         _edition: Edition,
         _cache: &mut Cache,
     ) -> Result<(Self, clean::Crate), Error> {
-        unimplemented!()
+        debug!("Initializing json renderer");
+        Ok((JsonRenderer { index: Rc::new(RefCell::new(FxHashMap::default())) }, krate))
     }
 
-    fn item(&mut self, _item: clean::Item, _cache: &Cache) -> Result<(), Error> {
-        unimplemented!()
+    fn item(&mut self, item: clean::Item, cache: &Cache) -> Result<(), Error> {
+        use clean::ItemEnum::*;
+        // Flatten items that recursively store other items by putting their children in the index
+        match item.inner.clone() {
+            StructItem(s) => s.fields.into_iter().for_each(|i| self.insert(i, cache)),
+            UnionItem(u) => u.fields.into_iter().for_each(|i| self.insert(i, cache)),
+            VariantItem(clean::Variant { kind: clean::VariantKind::Struct(v) }) => {
+                v.fields.into_iter().for_each(|i| self.insert(i, cache));
+            }
+            EnumItem(e) => e.variants.into_iter().for_each(|i| self.item(i, cache).unwrap()),
+            TraitItem(t) => t.items.into_iter().for_each(|i| self.insert(i, cache)),
+            ImplItem(i) => i.items.into_iter().for_each(|i| self.insert(i, cache)),
+            _ => {}
+        }
```

Comment:
I think it would be better to have this knowledge inside a method on `ItemEnum` itself, either returning an iterator of inner items or taking a closure.

Also, a problem with using a wildcard here is that a new kind of item with children could be added, and this match could be forgotten. So it's actually safer to list everything out.